election-server/src/db.rs

```diff
@@ -0,0 +1,310 @@
+            )
+            .await?;
+        let params: [&(dyn ToSql + Sync); 2] = [
+            &(contract_address.index as i64),
+            &(contract_address.subindex as i64),
+        ];
+        self.client.execute(&init_settings, &params).await?;
+        Ok(())
+    }
+
+    /// Get the latest block height recorded in the DB.
+    pub async fn get_settings(&self) -> DatabaseResult<StoredConfiguration> {
+        let get_settings = self
+            .client
+            .prepare_cached("SELECT latest_height, contract_index, contract_subindex FROM settings")
+            .await?;
+        self.client.query_one(&get_settings, &[]).await?.try_into()
+    }
+
+    /// Get ballot submission by transaction hash
+    pub async fn get_ballot_submission(
+        &self,
+        transaction_hash: TransactionHash,
+    ) -> DatabaseResult<Option<StoredBallotSubmission>> {
+        let get_ballot_submission = self
+            .client
+            .prepare_cached(
+                "SELECT transaction_hash, block_time, ballot, account, verified from ballots \
+                 WHERE transaction_hash = $1",
+            )
+            .await?;
+        let params: [&(dyn ToSql + Sync); 1] = [&transaction_hash.as_ref()];
+        let row = self
+            .client
+            .query_opt(&get_ballot_submission, &params)
+            .await?;
+        row.map(StoredBallotSubmission::try_from).transpose()
+    }
+
+    /// Get ballot submission by account address within the give range.
+    pub async fn get_ballot_submissions(
+        &self,
+        account_address: AccountAddress,
+        from: Option<usize>,
+        limit: usize,
+    ) -> DatabaseResult<Vec<StoredBallotSubmission>> {
+        let rows = if let Some(from) = from {
+            let get_ballot_submissions = self
+                .client
+                .prepare_cached(
+                    "SELECT id, transaction_hash, block_time, ballot, account, verified FROM \
+                     ballots WHERE account = $1 AND id < $2 ORDER BY id DESC LIMIT $3",
+                )
+                .await?;
+            let params: [&(dyn ToSql + Sync); 3] =
+                [&account_address.0.as_ref(), &(from as i64), &(limit as i64)];
+            self.client.query(&get_ballot_submissions, &params).await?
+        } else {
+            let get_ballot_submissions = self
+                .client
+                .prepare_cached(
+                    "SELECT id, transaction_hash, block_time, ballot, account, verified FROM \
+                     ballots WHERE account = $1 ORDER BY id DESC LIMIT $2",
+                )
+                .await?;
+            let params: [&(dyn ToSql + Sync); 2] = [&account_address.0.as_ref(), &(limit as i64)];
+            self.client.query(&get_ballot_submissions, &params).await?
+        };
+
+        rows.into_iter()
+            .map(StoredBallotSubmission::try_from)
+            .collect()
+    }
+}
+
+/// Wrapper around a database transaction
```

Comment:
Perhaps explain why it exists.

I'm also a bit confused as to why it has a (shared) reference to the transaction, rather than just wrapping the transction itself. It seems like one level of indirection too many.